virtual_ecosystem/models/abiotic_simple/microclimate.py

```diff
@@ -301,34 +271,28 @@ def interpolate_soil_temperature(
         soil temperature profile, [C]
     """
 
-    # select surface layer (atmosphere)
-    surface_layer = layer_heights[layer_heights.coords["layer_roles"] == "surface"]
-
-    # create array of interpolation heights including surface layer and soil layers
-    interpolation_heights = xr.concat(
-        [
-            surface_layer,
-            layer_heights[layer_heights.coords["layer_roles"] == "soil"] * -1
-            + surface_layer.values,
-        ],
-        dim="layers",
+    # select surface layer (atmosphere) and generate interpolation heights
+    surface_layer = layer_heights[layer_structure.role_indices["surface"]].to_numpy()
+    soil_depths = layer_heights[layer_structure.role_indices["all_soil"]].to_numpy()
+    interpolation_heights = np.concatenate(
+        [surface_layer, -1 * soil_depths + surface_layer]
     )
 
     # Calculate per cell slope and intercept for logarithmic soil temperature profile
-    slope = (surface_temperature - mean_annual_temperature) / (
-        np.log(interpolation_heights.isel(layers=0))
-        - np.log(interpolation_heights.isel(layers=-1))
+    slope = (surface_temperature.to_numpy() - mean_annual_temperature.to_numpy()) / (
+        np.log(interpolation_heights[0]) - np.log(interpolation_heights[-1])
     )
-    intercept = surface_temperature - slope * np.log(
-        interpolation_heights.isel(layers=0)
+    intercept = surface_temperature.to_numpy() - slope * np.log(
+        interpolation_heights[0]
     )
 
-    # Calculate the values within cells by layer
-    layer_values = np.log(interpolation_heights) * slope + intercept
+    # Calculate the values within cells by layer and clip by the bounds
+    layer_values = np.clip(
+        np.log(interpolation_heights) * slope + intercept, lower_bound, upper_bound
+    )
 
-    # set upper and lower bounds and return soil and surface layers, further layers are
-    # added in the 'run' function
-    return DataArray(
-        np.clip(layer_values, lower_bound, upper_bound),
-        coords=interpolation_heights.coords,
-    ).drop_isel(layers=0)
+    # return
+    return_xarray = layer_structure.from_template()
+    return_xarray[layer_structure.role_indices["all_soil"]] = layer_values[1:]
```

Comment:
It was cleaner to work internally with `numpy` arrays and then insert into the DataArray at the end.

Reply:
That's a good point, I noticed that when I tried to set up the climate_stats axis, too 👍